src/QsCompiler/DataStructures/Diagnostics.fs

```diff
@@ -544,7 +545,12 @@ type DiagnosticItem =
             | ErrorCode.ExpectingIterableExpr                     -> "The type {0} does not support iteration. Expecting an expression of array type or of type Range."
             | ErrorCode.ExpectingCallableExpr                     -> "The type of the expression must be a function or operation type. The given expression is of type {0}." 
             | ErrorCode.UnknownIdentifier                         -> "No identifier with the name \"{0}\" exists."
-                                                            
+            | ErrorCode.UnsupportedResultComparison               ->
+                // TODO: When the names of the runtime capabilities are finalized, they can be included in the error
+                // message.
+                "The execution target {0} does not support comparing measurement results. " +
+                "Choose an execution target with additional capabilities or use a controlled operation instead."
```

Comment:
I would get rid of this second half of the message - using controlled (a coherent if) is just usually not used with the same purpose as branching based on measurement result, and the suggestion hence seems odd to me.

Reply:
It can't be used in all cases (e.g. RUS loops), but `if (M(q1) == One) { X(q2); }` can be replaced with `Controlled X([q1], q2)`, for example.

Reply:
This came up in another thread, and Chris gave a really good rundown of why the controlled gates are actually quite different than conditionals:
https://github.com/microsoft/qsharp-runtime/pull/231#issuecomment-639585799
It may still be better than nothing in some scenarios, but it really can have rather major affects on overall system state.

Reply:
The intermediate state will be different - but I'm thinking of the [deferred measurement principle](https://en.wikipedia.org/wiki/Deferred_Measurement_Principle), which says that delaying measurements to the end of the circuit (using controlled gates) is equivalent to using classical controls with measurement.

Reply:
Oh, that’s interesting! Good point, and likely the best option for folks on platforms that don’t have conditional support, so I can see why you are recommending it. How does that affect what operations you can perform on the qubits in the interim between the controlled gate and the measurement? Just curious for my own edification, as I wonder what other limitations or concerns this pattern comes with.

Reply:
I still don't think having that part of the error message is the right thing to do. For one, as you say, it is not applicable in all situations - if it was Q# would be a whole lot less useful! Then, the people who understand it and will do the right thing don't need it, and for those who don't know that already I don't think the message is actionable since it does require some quantum understanding.

Reply:
How about:

> The execution target {0} does not support comparing measurement results. Choose an execution target with additional capabilities or avoid result comparisons.

My reasoning is: I want to mention the possibility of changing the execution target, so users know this limitation isn't inherent to Q#. I also want to mention the alternative (not comparing results), so that it doesn't look like the error message is saying the *only* solution is to change the execution target.

Reply:
> How does that affect what operations you can perform on the qubits in the interim between the controlled gate and the measurement? Just curious for my own edification, as I wonder what other limitations or concerns this pattern comes with.

I'm not aware of any limitations on the gates you can apply, but deferred measurement can increase the total number of qubits needed (measured qubits can be reused, but unmeasured qubits need to preserve their state for the whole circuit), and as mentioned loops aren't really possible. It can also increase errors since the unmeasured qubits need to stay coherent for longer.

Reply:
The message "The execution target {0} does not support comparing measurement results. Choose an execution target with additional capabilities or avoid result comparisons." sounds good to me.